src/suites/cts/validation/vertex_input.spec.ts

```diff
@@ -0,0 +1,619 @@
+    };
+    {
+      // Control case, setting different shader locations in two attributes
+      const descriptor = t.getDescriptor(vertexInput, VERTEX_SHADER_CODE_WITH_NO_INPUT);
+      t.device.createRenderPipeline(descriptor);
+    }
+    {
+      // Test same shader location in two attributes in the same buffer
+      vertexInput.vertexBuffers![0].attributeSet![1].shaderLocation = 0;
+      const descriptor = t.getDescriptor(vertexInput, VERTEX_SHADER_CODE_WITH_NO_INPUT);
+
+      await t.expectValidationError(() => {
+        t.device.createRenderPipeline(descriptor);
+      });
+    }
+  }
+  {
+    const vertexInput: GPUVertexInputDescriptor = {
+      vertexBuffers: [
+        {
+          stride: 0,
+          attributeSet: [
+            {
+              shaderLocation: 0,
+              format: 'float',
+            },
+          ],
+        },
+        {
+          stride: 0,
+          attributeSet: [
+            {
+              shaderLocation: 0,
+              format: 'float',
+            },
+          ],
+        },
+      ],
+    };
+    // Test same shader location in two attributes in different buffers
+    const descriptor = t.getDescriptor(vertexInput, VERTEX_SHADER_CODE_WITH_NO_INPUT);
+
+    await t.expectValidationError(() => {
+      t.device.createRenderPipeline(descriptor);
+    });
+  }
+});
+
+g.test('check out of bounds condition on attribute shader location', async t => {
+  const vertexInput: GPUVertexInputDescriptor = {
+    vertexBuffers: [
+      {
+        stride: 0,
+        attributeSet: [
+          {
+            shaderLocation: MAX_VERTEX_ATTRIBUTES - 1,
+            format: 'float',
+          },
+        ],
+      },
+    ],
+  };
+  {
+    // Control case, setting last attribute shader location
+    const descriptor = t.getDescriptor(vertexInput, VERTEX_SHADER_CODE_WITH_NO_INPUT);
+    t.device.createRenderPipeline(descriptor);
+  }
+  {
+    // Test attribute location OOB
+    vertexInput.vertexBuffers![0].attributeSet![0].shaderLocation = MAX_VERTEX_ATTRIBUTES;
+    const descriptor = t.getDescriptor(vertexInput, VERTEX_SHADER_CODE_WITH_NO_INPUT);
+
+    await t.expectValidationError(() => {
+      t.device.createRenderPipeline(descriptor);
+    });
+  }
+});
+
+g.test('check attribute offset out of bounds', async t => {
+  const vertexInput: GPUVertexInputDescriptor = {
+    vertexBuffers: [
+      {
+        stride: 0,
+        attributeSet: [
+          {
+            offset: MAX_VERTEX_BUFFER_END - Float32Array.BYTES_PER_ELEMENT,
+            shaderLocation: 0,
+            format: 'float',
+          },
+        ],
+      },
+    ],
+  };
+  {
+    // Control case, setting max attribute offset for Float32 vertex format
+    const descriptor = t.getDescriptor(vertexInput, VERTEX_SHADER_CODE_WITH_NO_INPUT);
+    t.device.createRenderPipeline(descriptor);
+  }
+  {
+    // Test attribute offset out of bounds
+    vertexInput.vertexBuffers![0].attributeSet![0].offset = MAX_VERTEX_BUFFER_END - 1;
```

Comment:
We also need a test that, for example, 3, 5, and 2045 *are* valid offsets. I don't think I saw one.

Reply:
Attribute offsets must be 4 byte aligned. This is why there were no tests for 3, 5, and 2045.

Reply:
I've added a test for "check multiple of 4 bytes constraint on offset" as it was missing